### ISSUE TITLE
Subpar inference results when exporting YOLOv8-cls model to ONNX

### ISSUE THREAD

**Issue by cisco-silva** [ultralytics/ultralytics#9964]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi! I am reaching out to discuss an issue I've encountered while utilizing the YOLOv8 model in ONNX format. Specifically, I'm observing a discrepancy in the inference results when employing the ONNX model directly via ONNXRuntime compared to using the Ultralytics YOLO wrapper. Despite using the same ONNX model in both scenarios, the outcomes are significantly different.

I have a custom-trained YOLOv8 model exported to ONNX format. When I perform inference using ONNXRuntime directly (referred to as Code 1 below), the results are substantially different from when I use the Ultralytics YOLO wrapper (referred to as Code 2 below). Notably, the confidence levels and predicted classes are inconsistent between the two methods. Just to be clear, the inference results from the YOLO wrapper (Code 2) are 100% aligned wit the ground truth of this set of images

**Code 1: Direct Inference using ONNXRuntime**
```
import onnxruntime
import numpy as np
import cv2
import os
import glob

class ClassificationModel:
    def __init__(self, model_path):
        self.model_path = model_path
        self.ort_session = onnxruntime.InferenceSession(self.model_path, providers=['CPUExecutionProvider'])
        self.input_shape = self.ort_session.get_inputs()[0].shape

    def preprocess(self, img_path):
        img = cv2.imread(img_path)
        img = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)
        
        height, width = img.shape[:2]
        scale = min(160 / height, 160 / width)
        
        new_height, new_width = int(height * scale), int(width * scale)
        resized_img = cv2.resize(img, (new_width, new_height))

        delta_w = 160 - new_width
        delta_h = 160 - new_height
        top, bottom = delta_h // 2, delta_h - (delta_h // 2)
        left, right = delta_w // 2, delta_w - (delta_w // 2)
        
        padded_img = cv2.copyMakeBorder(resized_img, top, bottom, left, right, cv2.BORDER_CONSTANT, value=[114, 114, 144])
        padded_img = padded_img.astype(np.float32) / 255.0
        padded_img = np.transpose(padded_img, (2, 0, 1))
        padded_img = np.expand_dims(padded_img, axis=0)

        return padded_img

    def main(self, img_path):
        preprocessed_img = self.preprocess(img_path)
        input_name = self.ort_session.get_inputs()[0].name
        outputs = self.ort_session.run(None, {input_name: preprocessed_img})
        return outputs[0][0]

model_path = "/home/models/best.onnx"
model = ClassificationModel(model_path)

imgs_path = "/home/dataset/test"
model_inputs = model.ort_session.get_modelmeta()
classes = eval(model_inputs.custom_metadata_map['names'])

for img_path in glob.glob(os.path.join(imgs_path, '*.jpg')):
    predictions = model.main(img_path)
    
    max_index = np.argmax(predictions)
    class_name = classes[max_index]
    confidence = predictions[max_index]  

    print(f"Image: {os.path.basename(img_path)} - Predicted class: {class_name}, Confidence: {confidence:.4f}")
```
**Code 2: Inference using Ultralytics YOLO Wrapper**
```
from ultralytics import YOLO

model = YOLO("/home/models/best.onnx")

source = "/home/dataset/test"

results = model(source, imgsz = 160, device = "CPU")  
```

**Results for Code 1:**
```
Image: 5601252231164_1_20231213_220207_649.jpg - Predicted class: hand_product, Confidence: 1.0000
Image: 0_2_20231225_131944_927.jpg - Predicted class: hand_product, Confidence: 0.7025
Image: 7622210641632_2_20240107_220130_284.jpg - Predicted class: hand_product, Confidence: 1.0000
Image: 7622210576095_1_20240107_213609_539.jpg - Predicted class: hand_product, Confidence: 1.0000
Image: 0_2_20240107_195151_632.jpg - Predicted class: hand_product, Confidence: 0.9995
Image: 7622210596413_1_20240107_213728_668.jpg - Predicted class: hand_product, Confidence: 0.9999
Image: 0_1_20240107_195141_999.jpg - Predicted class: nothing, Confidence: 0.7919
Image: 78936683_1_20231213_220356_760.jpg - Predicted class: hand_product, Confidence: 1.0000
Image: 0_1_20231225_131955_530.jpg - Predicted class: hand_product, Confidence: 0.9985
Image: 0_1_20231225_131938_142.jpg - Predicted class: hand_product, Confidence: 1.0000
```
**Results for Code 2:**
```
image 1/10 /home/dataset/test/0_1_20231225_131938_142.jpg: 160x160 nothing 1.00, hand_product 0.00, 
image 2/10 /home/dataset/test/0_1_20231225_131955_530.jpg: 160x160 nothing 1.00, hand_product 0.00, 
image 3/10 /home/dataset/test/0_1_20240107_195141_999.jpg: 160x160 nothing 1.00, hand_product 0.00, 
image 4/10 /home/dataset/test/0_2_20231225_131944_927.jpg: 160x160 nothing 1.00, hand_product 0.00, 
image 5/10 /home/dataset/test/0_2_20240107_195151_632.jpg: 160x160 nothing 1.00, hand_product 0.00, 
image 6/10 /home/dataset/test/5601252231164_1_20231213_220207_649.jpg: 160x160 hand_product 1.00, nothing 0.00, 
image 7/10 /home/dataset/test/7622210576095_1_20240107_213609_539.jpg: 160x160 hand_product 1.00, nothing 0.00, 
image 8/10 /home/dataset/test/7622210596413_1_20240107_213728_668.jpg: 160x160 hand_product 1.00, nothing 0.00, 
image 9/10 /home/dataset/test/7622210641632_2_20240107_220130_284.jpg: 160x160 hand_product 1.00, nothing 0.00, 
image 10/10 /home/dataset/test/78936683_1_20231213_220356_760.jpg: 160x160 hand_product 1.00, nothing 0.00, 
```

- I followed the recommended approach for using ONNX models with ONNXRuntime as per the official documentation.
- The ONNX model is verified to be correctly exported and functional.
- I would like to use the model with ONNXRuntime because of gain in inference speed.

I would appreciate any insights or guidance on why this discrepancy occurs and how I might resolve it to achieve consistent results across both methods.

**Comment by Kayzwer** [comment]
@cisco-silva try `model = YOLO("/home/models/best.onnx", "classify")`

**Comment by cisco-silva** [comment]
> @cisco-silva try `model = YOLO("/home/models/best.onnx", "classify")`

@Kayzwer thank you for your suggestion, but what I'm trying to achieve is that the ONNXRuntime (Code 1) implementation outputs the same results as the one in the YOLO wrapper (Code 2). My guess is that there is some preprocesssing step that the YOLO wrapper does that I'm not exactly replicating.

@glenn-jocher  Do you have any thoughts on this?

**Comment by glenn-jocher** [comment]
Hi @cisco-silva! 😊 It looks like you might need to ensure that your preprocessing in Code 1 matches exactly what the YOLO wrapper does. Typically, this involves not just resizing and normalization but also possibly handling the image format correctly (RGB vs. BGR). Here's a small tweak to your preprocessing to ensure it matches the expected input format:

```python
def preprocess(self, img_path):
    img = cv2.imread(img_path)
    img = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)  # Make sure to convert BGR to RGB

    # Your existing preprocessing code here...

    # Convert image to the RGB format (if not already done) and normalize
    padded_img = padded_img.astype(np.float32) / 255.0
    padded_img = np.transpose(padded_img, (2, 0, 1))
    padded_img = np.expand_dims(padded_img, axis=0)

    return padded_img
```

Also, make sure the input size (`imgsz`) and any other preprocessing steps you use align well with those used during training and by the YOLO wrapper. Sometimes, even small differences in preprocessing can lead to significant variances in inference outcomes. Hope this helps!

**Comment by glenn-jocher** [comment]
It seems likely that the discrepancy arises from differences in preprocessing. The Ultralytics YOLO wrapper includes specific preprocessing steps such as resizing, normalization, and padding, which must be replicated exactly in your ONNXRuntime code. I recommend inspecting the preprocessing implemented in the YOLO wrapper. You can refer to this [autobackend documentation](https://docs.ultralytics.com/reference/nn/autobackend/) for details on how inputs and outputs are handled, ensuring consistency between both methods. Let me know if further clarification is needed!

**Comment by keanuhouse** [comment]
I encountered the same problem as you. May I ask if you have found a solution? I would be very grateful if you could let me know.

**Comment by glenn-jocher** [comment]
@keanuhouse @cisco-silva The key issue is that the AutoBackend class in Ultralytics applies specific processing to both the input and output data when using ONNX models. 

Looking at your code, a few potential differences I notice:

1. In your padding values, you're using `[114, 114, 144]` which differs from the standard `[114, 114, 114]` used in Ultralytics
2. The classification outputs in Ultralytics are processed through a softmax function which I don't see in your direct ONNXRuntime implementation
3. There may be differences in how the outputs are interpreted - the YOLO wrapper might be applying additional post-processing

To investigate further, I recommend examining how the `autobackend.py` file handles output processing. For classification models specifically, there's additional handling to ensure proper normalization of confidence scores. You might need to apply softmax to your raw outputs to match the YOLO wrapper's behavior.

If you're looking for maximum inference speed while maintaining accuracy, you might also consider using the OpenVINO export format which often provides better CPU performance than standard ONNX.
